server/src/test/java/org/elasticsearch/action/support/master/info/TransportClusterInfoActionTests.java

```diff
@@ -0,0 +1,191 @@
+        clusterService = createClusterService(threadPool);
+        transportService = transport.createTransportService(clusterService.getSettings(), threadPool,
+            TransportService.NOOP_TRANSPORT_INTERCEPTOR, x -> clusterService.localNode(), null, Collections.emptySet());
+        transportService.start();
+        transportService.acceptIncomingRequests();
+        localNode = new DiscoveryNode("local_node", buildNewFakeTransportAddress(), Collections.emptyMap(),
+            Collections.singleton(DiscoveryNodeRole.MASTER_ROLE), Version.CURRENT);
+        allNodes = new DiscoveryNode[]{localNode};
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        clusterService.close();
+        transportService.close();
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+        threadPool = null;
+    }
+
+    static class Request extends ClusterInfoRequest<Request> {
+        Request() { }
+
+        Request(StreamInput in) throws IOException {
+            super(in);
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+            return new CancellableTask(id, type, action, "", parentTaskId, headers);
+        }
+    }
+
+    static class Response extends ActionResponse {
+        private long identity = randomLong();
+
+        Response() {}
+
+        Response(StreamInput in) throws IOException {
+            super(in);
+            identity = in.readLong();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeLong(identity);
+        }
+    }
+
+    static class Action extends TransportClusterInfoAction<Request, Response> {
+        Action(String actionName, TransportService transportService, ClusterService clusterService, ThreadPool threadPool) {
+            super(actionName,
+                transportService,
+                clusterService,
+                threadPool,
+                new ActionFilters(new HashSet<>()),
+                Request::new,
+                TestIndexNameExpressionResolver.newInstance(),
+                Response::new
+            );
+        }
+
+        @Override
+        protected void doMasterOperation(Task task,
+                                         Request request,
+                                         String[] concreteIndices,
+                                         ClusterState state,
+                                         ActionListener<Response> listener) {
+            listener.onResponse(new Response());
+        }
+    }
+
+    public void testGlobalBlocksAreCheckedBeforeResolvingIndices() throws Exception {
+        final boolean unblockBeforeTimeout = randomBoolean();
```

Comment:
Could we split this into two tests rather than randomly choosing between these branches?

Also (nit) I have a slight preference for making all the things `final`. Or none of them, but ideally not just the one.